File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/reporting/FormInfo.java
package com.urbanairship.android.layout.reporting;

import com.urbanairship.json.JsonMap;
import com.urbanairship.json.JsonSerializable;
import com.urbanairship.json.JsonValue;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class FormInfo implements JsonSerializable {
    @NonNull
    private final String identifier;

    @NonNull
    private final String formResponseType;

    @NonNull
    private final String formType;

    @Nullable
    private final Boolean isFormSubmitted;

    public FormInfo(@NonNull String identifier, @NonNull String formType, @Nullable String formResponseType, @Nullable Boolean isFormSubmitted) {
        this.identifier = identifier;
        this.formResponseType = formResponseType;
        this.formType = formType;
        this.isFormSubmitted = isFormSubmitted;
    }

    @NonNull
    public String getIdentifier() {
        return identifier;
    }

    @NonNull
    public String getFormResponseType() {
        return formResponseType;
    }

    @NonNull
    public String getFormType() {
        return formType;
    }

    @Nullable
    public Boolean getFormSubmitted() {
        return isFormSubmitted;
    }

    @Override
    public String toString() {
        return "FormInfo{" +
                "identifier='" + identifier + '\'' +
                ", formResponseType='" + formResponseType + '\'' +
                ", formType='" + formType + '\'' +
                ", isFormSubmitted=" + isFormSubmitted +
                '}';
    }

    @NonNull
    @Override
    public JsonValue toJsonValue() {
        return JsonMap.newBuilder()
                .put(KEY_IDENTIFIER, identifier)
                .put(KEY_FORM_RESPONSE_TYPE, formResponseType)
                .put(KEY_FORM_TYPE, formType)
                .putOpt(KEY_IS_FORM_SUBMITTED, isFormSubmitted)
                .build().toJsonValue();
    }

    private static final String KEY_IDENTIFIER = "identifier";
    private static final String KEY_FORM_RESPONSE_TYPE = "formResponseType";
    private static final String KEY_FORM_TYPE = "formType";
    private static final String KEY_IS_FORM_SUBMITTED = "isFormSubmitted";
}

File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/reporting/FormInfo.kt
package com.urbanairship.android.layout.reporting

import com.urbanairship.json.JsonMap.Companion.newBuilder
import com.urbanairship.json.JsonSerializable
import com.urbanairship.json.JsonValue
import com.urbanairship.json.jsonMapOf

public data class FormInfo public constructor(
    public val identifier: String,
    public val formType: String,
    public val formResponseType: String?,
    public val isFormSubmitted: Boolean?
) : JsonSerializable {

    override fun toJsonValue(): JsonValue = jsonMapOf(
        KEY_IDENTIFIER to identifier,
        KEY_FORM_RESPONSE_TYPE to formResponseType,
        KEY_FORM_TYPE to formType,
        KEY_IS_FORM_SUBMITTED to isFormSubmitted
    ).toJsonValue()

    private companion object {
        private const val KEY_IDENTIFIER = "identifier"
        private const val KEY_FORM_RESPONSE_TYPE = "formResponseType"
        private const val KEY_FORM_TYPE = "formType"
        private const val KEY_IS_FORM_SUBMITTED = "isFormSubmitted"
    }
}
